simp_le: 0.9.0 -> 0.16.0

picnoir opened the pull request:
##### Motivation for this change

As of today, let's encrypt dropped ACME V1 support. We need to update `simp_le` to `0.16.0` to support ACME V2.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @arianvp @flokli

picnoir committed:
simp_le: 0.9.0 -> 0.16.0

NinjaTrappeur commented:
Note: we need to update the let's encrypt service as well.

See discussion in https://github.com/NixOS/nixpkgs/issues/70966

arianvp commented:
@GrahamcOfBorg test acme

NinjaTrappeur commented:
Trying to update `acme` as well.

There's something wrong with the acme-ci integration test setup:

```
    workers = ['primary'] if not config.option.numprocesses\
E   AttributeError: 'Namespace' object has no attribute 'numprocesses'
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
```

Coming from there: https://github.com/certbot/certbot/blob/master/certbot-ci/certbot_integration_tests/conftest.py#L85

Investigating.

NinjaTrappeur commented:
Should I disable this test suite (`acme-ci`) altogether?

NinjaTrappeur commented:
I have to go offline for today.

Since this is a somehow critical situation (letsencrypt is broken), I disabled the certbot tests altogether.

In order to re-enable them, we have to pass the `--numprocesses 1` flag to `pytest`.

I am not familiar enough with the nixos python infrastructure to know how to do it and did not manage to track down were this override should happen.

I'll have a second look to this issue that tomorrow. In the meantime, feel free to either merge this PR with the disabled tests or edit it to pass the appropriate flag to the test runner.

WilliButz reviewed (changes requested):
As mentioned in the [changelog](https://github.com/zenhack/simp_le/blob/master/README.rst#0150), this version does not work out of the box with the previous configuration, so the module needs an update as well.

Adding the required flag did the trick for me, but I'd like someone else who is more familiar with the whole acme module to do a review before merging :)
```diff
diff --git a/nixos/modules/security/acme.nix b/nixos/modules/security/acme.nix
index b321c04e574..e35ea0c7b92 100644
--- a/nixos/modules/security/acme.nix
+++ b/nixos/modules/security/acme.nix
@@ -69,9 +69,9 @@ let
       plugins = mkOption {
         type = types.listOf (types.enum [
           "cert.der" "cert.pem" "chain.pem" "external.sh"
-          "fullchain.pem" "full.pem" "key.der" "key.pem" "account_key.json"
+          "fullchain.pem" "full.pem" "key.der" "key.pem" "account_key.json" "account_reg.json"
         ]);
-        default = [ "fullchain.pem" "full.pem" "key.pem" "account_key.json" ];
+        default = [ "fullchain.pem" "full.pem" "key.pem" "account_key.json" "account_reg.json" ];
         description = ''
           Plugins to enable. With default settings simp_le will
           store public certificate bundle in <filename>fullchain.pem</filename>,
```

NinjaTrappeur commented:
Thanks @dtzWill,

Just updated with your requested change.

The nixos test still fails though:

```
acmeStandalone# [   19.777060] simp_le[737]: Traceback (most recent call last):
acmeStandalone# [   19.779146] simp_le[737]:   File "/nix/store/dj30chq7dlnh8aphj1bq79b4hjlmj47p-python3.7-urllib3-1.24.3/lib/python3.7/site-packages/urllib3/connection.py", line 159, in _new_conn
acmeStandalone# [   19.787745] simp_le[737]:     (self._dns_host, self.port), self.timeout, **extra_kw)
acmeStandalone# [   19.789733] simp_le[737]:   File "/nix/store/dj30chq7dlnh8aphj1bq79b4hjlmj47p-python3.7-urllib3-1.24.3/lib/python3.7/site-packages/urllib3/util/connection.py", line 57, in create_connection
acmeStandalone# [   19.797023] simp_le[737]:     for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
acmeStandalone# [   19.804943] simp_le[737]:   File "/nix/store/8lhmyjarm73453f2mdz0xli9w8sy0wvh-python3-3.7.4/lib/python3.7/socket.py", line 748, in getaddrinfo
acmeStandalone# [   19.812701] simp_le[737]:     for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
acmeStandalone# [   19.819733] simp_le[737]: socket.gaierror: [Errno -2] Name or service not known
acmeStandalone# [   19.821687] simp_le[737]: During handling of the above exception, another exception occurred:
acmeStandalone# [   19.828823] simp_le[737]: Traceback (most recent call last):
acmeStandalone# [   19.830757] simp_le[737]:   File "/nix/store/dj30chq7dlnh8aphj1bq79b4hjlmj47p-python3.7-urllib3-1.24.3/lib/python3.7/site-packages/urllib3/connectionpool.py", line 600, in urlopen
acmeStandalone# [   19.838018] simp_le[737]:     chunked=chunked)
acmeStandalone# [   19.844801] simp_le[737]:   File "/nix/store/dj30chq7dlnh8aphj1bq79b4hjlmj47p-python3.7-urllib3-1.24.3/lib/python3.7/site-packages/urllib3/connectionpool.py", line 343, in _make_request
acmeStandalone# [   19.852721] simp_le[737]:     self._validate_conn(conn)
acmeStandalone# [   19.854093] simp_le[737]:   File "/nix/store/dj30chq7dlnh8aphj1bq79b4hjlmj47p-python3.7-urllib3-1.24.3/lib/python3.7/site-packages/urllib3/connectionpool.py", line 839, in _validate_conn
acmeStandalone# [   19.860752] simp_le[737]:     conn.connect()
acmeStandalone# [   19.861986] simp_le[737]:   File "/nix/store/dj30chq7dlnh8aphj1bq79b4hjlmj47p-python3.7-urllib3-1.24.3/lib/python3.7/site-packages/urllib3/connection.py", line 301, in connect
acmeStandalone# [   19.869857] simp_le[737]:     conn = self._new_conn()
acmeStandalone# [   19.871853] simp_le[737]:   File "/nix/store/dj30chq7dlnh8aphj1bq79b4hjlmj47p-python3.7-urllib3-1.24.3/lib/python3.7/site-packages/urllib3/connection.py", line 168, in _new_conn
acmeStandalone# [   19.883110] simp_le[737]:     self, "Failed to establish a new connection: %s" % e)
acmeStandalone# [   19.890968] simp_le[737]: urllib3.exceptions.NewConnectionError: <urllib3.connection.VerifiedHTTPSConnection object at 0x7fcc238c6f50>: Failed to establish a new connection: [Errno -2] Name or service not known
acmeStandalone# [   19.899746] simp_le[737]: During handling of the above exception, another exception occurred:
acmeStandalone# [   19.901795] simp_le[737]: Traceback (most recent call last):
acmeStandalone# [   19.904480] simp_le[737]:   File "/nix/store/k27rv5djq90f93rl28b08aj493061lcz-python3.7-requests-2.22.0/lib/python3.7/site-packages/requests/adapters.py", line 449, in send
acmeStandalone# [   19.914962] simp_le[737]:     timeout=timeout
acmeStandalone# [   19.917055] simp_le[737]:   File "/nix/store/dj30chq7dlnh8aphj1bq79b4hjlmj47p-python3.7-urllib3-1.24.3/lib/python3.7/site-packages/urllib3/connectionpool.py", line 638, in urlopen
acmeStandalone# [   19.923495] simp_le[737]:     _stacktrace=sys.exc_info()[2])
acmeStandalone# [   19.926365] simp_le[737]:   File "/nix/store/dj30chq7dlnh8aphj1bq79b4hjlmj47p-python3.7-urllib3-1.24.3/lib/python3.7/site-packages/urllib3/util/retry.py", line 399, in increment
acmeStandalone# [   19.930136] simp_le[737]:     raise MaxRetryError(_pool, url, error or ResponseError(cause))
acmeStandalone# [   19.933354] simp_le[737]: urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='acme-v02.api.letsencrypt.org', port=443): Max retries exceeded with url: /directory (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fcc238c6f50>: Failed to establish a new connection: [Errno -2] Name or service not known'))
acmeStandalone# [   19.946894] simp_le[737]: During handling of the above exception, another exception occurred:
acmeStandalone# [   19.950181] simp_le[737]: Traceback (most recent call last):
acmeStandalone# [   19.954652] simp_le[737]:   File "/nix/store/ddgffgv61bpirahy4sxdg53703izd45j-simp_le-client-0.16.0/lib/python3.7/site-packages/simp_le.py", line 1565, in main
acmeStandalone# [   19.960103] simp_le[737]:     return main_with_exceptions(cli_args)
acmeStandalone# [   19.964639] simp_le[737]:   File "/nix/store/ddgffgv61bpirahy4sxdg53703izd45j-simp_le-client-0.16.0/lib/python3.7/site-packages/simp_le.py", line 1549, in main_with_exceptions
acmeStandalone# [   19.973666] simp_le[737]:     persist_new_data(args, existing_data)
acmeStandalone# Job for acme-standalone.com.service failed because the control process exited with error code.
acmeStandalone# See "systemctl status acme-standalone.com.service" and "journalctl -xe" for details.[   19.984113]
acmeStandalone# simp_le[737]:   File "/nix/store/ddgffgv61bpirahy4sxdg53703izd45j-simp_le-client-0.16.0/lib/python3.7/site-packages/simp_le.py", line 1408, in persist_new_data
acmeStandalone# [   19.988593] simp_le[737]:     args, existing_data.account_key, existing_data.account_reg)
acmeStandalone# [   19.992414] simp_le[737]:   File "/nix/store/ddgffgv61bpirahy4sxdg53703izd45j-simp_le-client-0.16.0/lib/python3.7/site-packages/simp_le.py", line 1341, in registered_client
acmeStandalone# [   19.996954] simp_le[737]:     directory = messages.Directory.from_json(net.get(args.server).json())
acmeStandalone# [   20.003102] simp_le[737]:   File "/nix/store/5703sd0w9h9gk3ii51fc6kkwhbadaiv8-python3.7-acme-0.39.0/lib/python3.7/site-packages/acme/client.py", line 1161, in get
acmeStandalone# [   20.008666] simp_le[737]:     self._send_request('GET', url, **kwargs), content_type=content_type)
acmeStandalone# [   20.012744] systemd[1]: acme-standalone.com.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
```

Will look into it in the next ~4 hours. In the meantime, if somebody familiar with this whole area understands what's happening, feel free to edit this PR.

NinjaTrappeur commented:
Updated nixos tests to use acme-V02 endpoints.

We are currently using an outdated `boulder` version for the tests; which is not compatible with the `acme-v02` protocol.

We probably want:

1. Extract boulder from the nixos test and declare it as a regular nixpkgs pkg.
2. Update the boulder derivation to `release-2019-10-13`

I won't have enough spare time to do that today, would someone be willing to share the migration effort here?

adisbladis committed:
boulder: init at release-2019-10-13

adisbladis commented:
@NinjaTrappeur I took a few minutes and packaged up Boulder in https://github.com/NixOS/nixpkgs/pull/71324

manveru commented:
FWIW, i'm using this with the fix from @WilliButz and it works fine :)

flokli commented:
> FWIW, i'm using this with the fix from @WilliButz and it works fine :)

What's the fix? Can you link to it? ;-)

manveru commented:
@flokli I made a branch at https://github.com/manveru/nixpkgs/commit/956169b37546dafb90996330306415d7238249a4

NinjaTrappeur commented:
@flokli @manveru It's included in this PR right? Am I missing something?

jtojnar commented on pkgs/tools/admin/simp_le/default.nix:
```diff
@@ -2,16 +2,16 @@
   postPatch = ''
-    # drop upper bound of acme requirement
-    sed -ri "s/'(acme>=[^,]+),<[^']+'/'\1'/" setup.py
+    # drop acme bounds
```
Should not we update (certbot) as https://github.com/NixOS/nixpkgs/pull/62283? Upstream must have bumped the lower bound for a reason.

NinjaTrappeur replied:
You're right. I did that in an attempt to quickly have something working. On it.

NinjaTrappeur replied:
Fixed, thanks for the heads up.

jtojnar commented on pkgs/tools/admin/certbot/default.nix:
```diff
@@ -2,25 +2,27 @@
     psutil
     pyRFC3339
     pyopenssl
+    pytest
```
Shouldn't this go to `checkInputs`?

aszlig commented:
Just wanted to chime in on this, there is also [Pebble](https://github.com/letsencrypt/pebble). Back then when writing the initial implementation of the ACME test, Pebble was pretty much WIP. Since some time has passed, maybe it makes sense to re-evaluate and use Pebble instead, because it should reduce much of the complexity involved.

picnoir committed:
certbot: 0.31.0 -> 0.39.0

Updating:

- nixos module to use the new `account_reg.json` file.
- use nixpkgs pebble for integration tests.

Co-authored-by: Florian Klink <flokli@flokli.de>

Replace certbot-embedded pebble

flokli commented:
@aszlig `certbot` uses `pebble` for their integration tests. However, they also seem to require network access - see https://github.com/certbot/certbot/issues/7450.

picnoir committed:
nixos/tests/letsencrypt: use Pebble instead of Boulder

Let's encrypt bumped ACME to V2. We need to update our nixos test to
be compatible with this new protocol version.

We decided to drop the Boulder ACME server in favor of the more
integration test friendly Pebble.

- overriding cacert not necessary
- this avoids rebuilding lots of packages needlessly
- nixos/tests/acme: use pebble's ca for client tests
- pebble always generates its own ca which has to be fetched

TODO: write proper commit msg :)

NinjaTrappeur commented:
> @aszlig `certbot` uses `pebble` for their integration tests. However, they also seem to require network access - see https://github.com/certbot/certbot/issues/7450.

I think we are talking about replacing the whole boulder dance in the
nixos test with the simpler-to-deploy `pebble`.

I think this is a good idea.

However I'd like first to focus on having the current setup up and
running given the tight timeframe regarding the ACME V2 deployment.

Maybe we should open an issue about this refactoring? WDYT

c0bw3b commented on pkgs/tools/admin/certbot/default.nix:
```diff
@@ -40,7 +71,15 @@ python3Packages.buildPythonApplication rec {
-  doCheck = !stdenv.isDarwin; # On Hydra Darwin tests fail with "Too many open files".
+  # tests currently time out, because they're trying to do network access
+  # Upstream issue: https://github.com/certbot/certbot/issues/7450
+  doCheck = false
```
```suggestion
  doCheck = false;
```

NinjaTrappeur commented:
Current status:

- [x] Simple updated
- [x] Boulder updated (courtesy of Adisbadis)
- [x] Certbot updated (courtesy of Flokli)
- [ ] Adapting the nixos test - WIP.

If updating the test turns out to be a nightmare, I might switch to `pebble` as suggested by Azlig.

jtojnar commented on nixos/tests/common/letsencrypt/default.nix:
```diff
@@ -135,47 +135,15 @@ let
-      rm -r go/src/github.com/letsencrypt/boulder/vendor/github.com/miekg/pkcs11
-    '';
-
-    # XXX: Temporarily brought back putting the source code in the output,
```
Do not remove the explanation. Or maybe change it to the first paragraph of https://github.com/NixOS/nixpkgs/commit/c21b1ede95de8bead6a83b8f182fd17f2c6a0ee0 to be even clearer.

arianvp commented:
@GrahamcOfBorg test acme

NinjaTrappeur commented:
Note: the acme nixos test is currently trying to build `linux-4.19.79`.

This linux dependency seems to be introduced by `/xbk6fg6k74jrdnr4iaidn319vmdjvxgp-nixos-system-letsencrypt-20.03.git.d2f432e.drv`. That said, I did not manage to understand why and where yesterday night :/

flokli commented:
Rebased to latest master.

@GrahamcOfBorg build nixosTests.acme

flokli commented:
I assume the rebuilds are mostly triggered due to changes in master not yet in the binary caches. I built locally, however it seems the acme test still fails:

```
letsencrypt: waiting for unit ‘boulder.service’
letsencrypt: running command: systemctl --no-pager show "boulder.service"
letsencrypt# [   20.132153] boulder-wfe[1054]: E130517 boulder-wfe [AUDIT] grpc: addrConn.createTransport failed to connect to {127.0.0.1:9095 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:9095: connect: connection refused". Reconnecting...
letsencrypt# [   20.142004] boulder-ra[1051]: [AUDIT] grpc: addrConn.createTransport failed to connect to {127.0.0.1:9095 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:9095: connect: connection refused". Reconnecting...
letsencrypt# [   20.150071] boulder-ra[1051]: E130517 boulder-ra [AUDIT] grpc: addrConn.createTransport failed to connect to {127.0.0.1:9095 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:9095: connect: connection refused". Reconnecting...
letsencrypt# [   20.162620] boulder-ra[1051]: E130517 boulder-ra [AUDIT] grpc: failed dns A record lookup due to lookup akamai-purger.boulder: no such host.
letsencrypt# [   20.171716] boulder-ra[1051]: 
letsencrypt# [   20.198306] boulder-va[1053]: I130517 boulder-va Versions: boulder-va=(Unspecified Unspecified) Golang=(go1.13.1) BuildHost=(Unspecified)
letsencrypt# [   20.235535] systemd[1]: boulder-ct-test-srv.service: Service RestartSec=100ms expired, scheduling restart.
letsencrypt# [   20.262763] boulder-ca[1050]: I130517 boulder-ca Versions: boulder-ca=(Unspecified Unspecified) Golang=(go1.13.1) BuildHost=(Unspecified)
letsencrypt# [   20.280948] boulder-ca[1050]: I130517 boulder-ca loading hostname policy, sha256: 66fb49722da7f9923f9d0c3aaebd87aea4a3c35457ee8b5fae59ebdc32d140dc
letsencrypt# [   20.292783] boulder-ca[1050]: E130517 boulder-ca [AUDIT] Couldn't load private key: pkcs11key: problem making Key: unable to load PKCS#11 module from "/usr/local/lib/libpkcs11-proxy.so"
letsencrypt: exit status 0
(0.23 seconds)
letsencrypt: running command: systemctl list-jobs --full 2>&1
letsencrypt# [   20.307738] boulder-publisher[1058]: E130517 boulder-publisher [AUDIT] grpc: Server.Serve failed to complete security handshake from "127.0.0.1:57344": EOF
letsencrypt# [   20.360670] l946djg27j24q41xxbiglglbb8d5a2y2-unit-script-boulder-boulder-publisher-start[1059]: Connection to 127.0.0.1 9091 port [tcp/xmltec-xmlmail] succeeded!
letsencrypt# [   20.391749] boulder-wfe[1054]: [AUDIT] grpc: addrConn.createTransport failed to connect to {127.0.0.1:9095 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:9095: connect: connection refused". Reconnecting...
letsencrypt: exit status 0
(0.06 seconds)
letsencrypt: running command: systemctl --no-pager show "boulder.service"
letsencrypt# [   20.408730] boulder-ra[1051]: E130517 boulder-ra [AUDIT] grpc: addrConn.createTransport failed to connect to {127.0.0.1:9093 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:9093: connect: connection refused". Reconnecting...
letsencrypt# [   20.416146] systemd[1]: boulder-ct-test-srv.service: Scheduled restart job, restart counter is at 1.
letsencrypt# [   20.424939] boulder-ca[1050]: Couldn't load private key: pkcs11key: problem making Key: unable to load PKCS#11 module from "/usr/local/lib/libpkcs11-proxy.so"
letsencrypt# [   20.432282] boulder-ra[1051]: [AUDIT] grpc: failed dns A record lookup due to lookup akamai-purger.boulder: no such host.
letsencrypt# [   20.437100] systemd[1]: Stopped Boulder ACME Component (ct-test-srv).
letsencrypt# [   20.444973] boulder-publisher[1058]: [AUDIT] grpc: Server.Serve failed to complete security handshake from "127.0.0.1:57344": EOF
letsencrypt# [   20.459699] systemd[1]: Started Boulder ACME Component (ct-test-srv).
letsencrypt# [   20.478797] boulder-ca[1050]: [AUDIT] Couldn't load private key: pkcs11key: problem making Key: unable to load PKCS#11 module from "/usr/local/lib/libpkcs11-proxy.so"
letsencrypt# [   20.492048] systemd[1]: Started Wait For Port 9091 (boulder-publisher).
letsencrypt# [   20.500577] ct-test-srv[1172]: 2019/10/20 13:05:18 open : no such file or directory
letsencrypt# [   20.508170] boulder-ra[1051]: [AUDIT] grpc: addrConn.createTransport failed to connect to {127.0.0.1:9093 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:9093: connect: connection refused". Reconnecting...
letsencrypt# [   20.520235] systemd[1]: Started Boulder ACME Component (ocsp-updater).
letsencrypt# [   20.527130] systemd[1]: Startup finished in 4.788s (kernel) + 15.260s (userspace) = 20.049s.
```

There seems to be multiple issues, `builder-ca` trying to load `/usr/local/lib/libpkcs11-proxy.so`, `boulder-ra` trying to lookup `akamai-purger.boulder`, `boulder-ra` trying to connect to `127.0.0.1:9093`, `boulder-wfe` trying to connect to `127.0.0.1:9095`…

flokli commented:
Seems `pebble` might be much easier to use in the test - we'll give it a try. Thanks @aszlig for the pointer!

jtojnar commented:
> I assume the rebuilds are mostly triggered due to changes in master not yet in the binary caches.

There must be something different in play. The rebuild is still triggered when the four commits are cherry-picked onto nixos-unstable, but on just unstable, everything is cached.

NinjaTrappeur commented:
> There must be something different in play. The rebuild is still triggered when the four commits are cherry-picked onto nixos-unstable, but on just unstable, everything is cached.

Yup. That's why I'm puzzled.

> Seems pebble might be much easier to use in the test - we'll give it a try. Thanks @aszlig for the pointer!

+1. Will have a go to this tonight (CEST).

risicle commented:
This implicitly breaks `certbot` on macos.

picnoir committed:
pebble: init at 2.2.2

NinjaTrappeur commented:
Status: I updated the nixos test to use `pebble` instead of `boulder`. This greatly simplifies the test setup.

The test fails to verify the `ACME` challenge though:

```
webserver# [   57.835630] simp_le[958]: 2019-10-21 09:18:24,315:DEBUG:acme.client:1074: Ignoring wrong Content-Type ('application/json; charset=utf-8') for JSON decodable response
webserver# [   57.837970] simp_le[958]: 2019-10-21 09:18:24,315:DEBUG:acme.client:1170: Storing nonce: _FOm_wjBIrYH5Ksz_Hh_NA
webserver# [   57.840068] simp_le[958]: 2019-10-21 09:18:24,315:ERROR:simp_le:1396: CA marked some of the authorizations as invalid, which likely means it could not access http://example.com/.well-known/acme-challenge/X. Did you set correct path in -d example.com:path or --default_root? Are all your domains accessible from the internet? Please check your domains' DNS entries, your host's network/firewall setup and your webserver config. If a domain's DNS entry has both A and AAAA fields set up, some CAs such as Let's Encrypt will perform the challenge validation over IPv6. If your DNS provider does not answer correctly to CAA records request, Let's Encrypt won't issue a certificate for your domain (see https://letsencrypt.org/docs/caa/). Failing authorizations: https://acme-v02.api.letsencrypt.org/authZ/opP7vohupK_49LzEKh62LXnMGTcd-meB-WkMTG7g7FA
webserver# [   57.849415] simp_le[958]: 2019-10-21 09:18:24,315:DEBUG:simp_le:1098: Removing validation file at /var/lib/acme/acme-challenge/.well-known/acme-challenge/JrPTTeuRm-JZnHXl4mlFw_zDjFtc6FHQljSxVLJvHAc
webserver# [   57.852355] simp_le[958]: Traceback (most recent call last):
webserver# [   57.853799] simp_le[958]:   File "/nix/store/na49jaw0ym7mr9cvfg1wya0izav0dxn7-simp_le-client-0.16.0/lib/python3.7/site-packages/simp_le.py", line 1370, in finalize_order
webserver# [   57.856380] simp_le[958]:     finalized_order = client.poll_and_finalize(order)
webserver# [   57.864375] simp_le[958]:   File "/nix/store/dprgpwvqjrgmmzkj9z0zslajqm646qs3-python3.7-acme-0.39.0/lib/python3.7/site-packages/acme/client.py", line 712, in poll_and_finalize
webserver# [   57.867797] simp_le[958]:     orderr = self.poll_authorizations(orderr, deadline)
webserver# [   57.870065] simp_le[958]:   File "/nix/store/dprgpwvqjrgmmzkj9z0zslajqm646qs3-python3.7-acme-0.39.0/lib/python3.7/site-packages/acme/client.py", line 736, in poll_authorizations
webserver# [   57.875439] simp_le[958]:     raise errors.ValidationError(failed)
webserver# [   57.877772] simp_le[958]: acme.errors.ValidationError
webserver# [   57.880609] simp_le[958]: During handling of the above exception, another exception occurred:
webserver# [   57.883472] simp_le[958]: Traceback (most recent call last):
webserver# [   57.885852] simp_le[958]:   File "/nix/store/na49jaw0ym7mr9cvfg1wya0izav0dxn7-simp_le-client-0.16.0/lib/python3.7/site-packages/simp_le.py", line 1443, in persist_new_data
webserver# [   57.888833] simp_le[958]:     order = finalize_order(client, order)
webserver# [   57.890910] simp_le[958]:   File "/nix/store/na49jaw0ym7mr9cvfg1wya0izav0dxn7-simp_le-client-0.16.0/lib/python3.7/site-packages/simp_le.py", line 1397, in finalize_order
webserver# [   57.893932] simp_le[958]:     raise Error('Challenge validation has failed, see error log.')
webserver# [   57.896218] simp_le[958]: simp_le.Error: Challenge validation has failed, see error log.
webserver# [   57.898495] simp_le[958]: During handling of the above exception, another exception occurred:
webserver# [   57.900684] simp_le[958]: Traceback (most recent call last):
webserver# [   57.910045] simp_le[958]:   File "/nix/store/na49jaw0ym7mr9cvfg1wya0izav0dxn7-simp_le-client-0.16.0/lib/python3.7/site-packages/simp_le.py", line 1565, in main
webserver# [   57.913206] simp_le[958]:     return main_with_exceptions(cli_args)
webserver# [   57.915430] simp_le[958]:   File "/nix/store/na49jaw0ym7mr9cvfg1wya0izav0dxn7-simp_le-client-0.16.0/lib/python3.7/site-packages/simp_le.py", line 1549, in main_with_exceptions
webserver# [   57.918763] simp_le[958]:     persist_new_data(args, existing_data)
webserver# [   57.925412] simp_le[958]:   File "/nix/store/na49jaw0ym7mr9cvfg1wya0izav0dxn7-simp_le-client-0.16.0/lib/python3.7/site-packages/simp_le.py", line 1464, in persist_new_data
webserver# [   57.934429] simp_le[958]:     chain=None,
webserver# [   57.942299] simp_le[958]:   File "/nix/store/na49jaw0ym7mr9cvfg1wya0izav0dxn7-simp_le-client-0.16.0/lib/python3.7/site-packages/simp_le.py", line 1128, in persist_data
webserver# [   57.955529] simp_le[958]:     plugin.save(new_data)
webserver# [   57.956973] simp_le[958]:   File "/nix/store/na49jaw0ym7mr9cvfg1wya0izav0dxn7-simp_le-client-0.16.0/lib/python3.7/site-packages/simp_le.py", line 648, in savethe following new units were started:
webserver# [   57.965594] simp_le[958]:     pems = [self.dump_cert(data.cert)]
webserver# [   57.967550] simp_le[958]:   File "/nix/store/na49jaw0ym7mr9cvfg1wya0izav0dxn7-simp_le-client-0.16.0/lib/python3.7/site-packages/simp_le.py", line 468, in dump_certacme-b.example.com.service, acme-b.example.com.timer, acme-finished-b.example.com.target
webserver# warning: the following units failed: acme-a.example.com.service
```

I won't have any spare time to dedicate to this for the next couple of days, feel free to push on this PR if needed.

NinjaTrappeur commented on pkgs/tools/admin/certbot/default.nix:
```diff
@@ -1,21 +1,44 @@
 
-python3Packages.buildPythonApplication rec {
+let
+  pebble_linux_amd64 = stdenv.mkDerivation rec {
```
Note to self: reuse `pkgs.pebble` here.

fpletz commented:
Fixed the test and made the tests run faster without rebuilding so many packages. See the messages in the individual commits.

TODO: Squash the commits together. Will do that tomorrow if nobody beats me to it.

fpletz commented:
@GrahamcOfBorg test acme

fpletz commented:
@GrahamcOfBorg build nixosTests.acme

aszlig commented on nixos/tests/common/letsencrypt/common.nix:
```diff
@@ -1,12 +1,17 @@
+
   nixpkgs.overlays = lib.singleton (self: super: {
-    cacert = super.cacert.overrideDerivation (drv: {
+    cacert-le = super.cacert.overrideDerivation (drv: {
```
```suggestion
    cacert-le = super.cacert-le.overrideDerivation (drv: {
```

aszlig replied:
While this does trim down the rebuilds, the reason I made an override on `cacert` in the first place was to make sure that we override as many variants as possible (eg. all packages that use `cacert` in any way) because the module is designed to be generally used not only for `tests/acme.nix`, that's why the module is in `common/`.

aszlig replied:
To elaborate more on that: It's designed in a way to be used as a drop-in module you can use for writing full-network tests that should be as close as possible to the real world, eg. you just throw in `common/resolver.nix` and `common/letsencrypt` and can use the production config of your deployment for testing, without the need to patch in snakeoil certs.

For example, in a few deployments I don't use the ACME module from NixOS, because I need DNS-01 challenges and so can test the implementation by reusing the modules from `tests/common` without doing a lot of extra patching.

Btw., another option here to avoid the rebuilds could be `system.replaceRuntimeDependencies`.

flokli replied:
@aszlig while a lot of packages transitively depend on `cacert` during build time, they normally don't bake the certificate store inside their binaries, but we point them to look into `/etc/ssl/certs/ca-(certificates,bundle).crt` at runtime. Custom CAs can be added there through the `security.pki.certificate[Files]` options, which should not rebuild the world, and I'd consider a software not honoring this location a bug.

NinjaTrappeur replied:
I can update the PR to use that later today!

Thanks for the help @fpletz btw, it's very much appreciated :)

fpletz replied:
@aszlig I agree in principle but as far as I understood pebble's code, you can't provide a CA to be used to issue certificates but rather pebble seems to generate its own temporary CA which can be fetched via HTTP after it's started. So we don't know the CA beforehand and thus we can't even override `pkgs.cacert` or amend `/etc/ssl/certs/ca-certs.crt`.

`pythonPackages.certifi` is still overridden because it has it's own CA trust store (I agree with @flokli that this is technically a bug) that is used by `simp_le`. The snakeoil CA and cert is only used to verify requests to `acme-v02.api.letsencrypt.org` but nothing else.

aszlig replied:
@fpletz: Hm, maybe it makes sense to stick with Boulder then :-/ Or alternatively patch Pebble to use a pre-generated root/intermediate.

@NinjaTrappeur: If you want I can look into getting Boulder to work tonight.

flokli replied:
@aszlig The tests now pass, we download the certificates from pebble and manually specify them while running curl. While this requires some custom setup in the `testScript`, I think it's still wayy less complicated compared to setting up boulder (or patching pebble to use a pre-generated CA - it's only kept in memory currently)

I'm a bit worried about boulder being a bit too complicated just for a NixOS test - if it would be a proper NixOS module, with someone actually using it in a production setting, that would be another thing, but at least for now, I'd prefer sticking with manually fetching and setting `--cacert` instead of having a large amount of test harness code.

Are you fine with that?

aszlig replied:
@flokli: I think getting in ACMEv2 support is much more important right now rather than being able to use the test module for full-network tests, given that I might (we don't have metrics, so just a guess) be the only one actually using it.

So yes, I'm fine with that. We can still fix the full-network case later.

NinjaTrappeur commented:
Just updated the PR according to both @aszlig and @flokli remarks.

The test fails with:

```
letsencrypt# [   55.767541] mzplx3xcy4g0qv1i60416swxb6bx3g4k-unit-script-pebble-start[437]: 2019/10/23 11:01:46 http: TLS handshake error from 192.168.1.4:56988: local error: tls: bad record MAC
webserver# [   28.167422] simp_le[924]: Traceback (most recent call last):
webserver# [   28.168924] simp_le[924]:   File "/nix/store/84r6h04cxn5yqd7za82d494rkr24rx91-python3.7-urllib3-1.24.3/lib/python3.7/site-packages/urllib3/contrib/pyopenssl.py", line 456, in wrap_socket
webserver# [   28.178282] simp_le[924]:     cnx.do_handshake()
webserver# [   28.179704] simp_le[924]:   File "/nix/store/xxg8w24n7nklwc6lcldvc3j7xwv4bfmr-python3.7-pyOpenSSL-19.0.0/lib/python3.7/site-packages/OpenSSL/SSL.py", line 1915, in do_handshake
webserver# [   28.191301] simp_le[924]:     self._raise_ssl_error(self._ssl, result)
webserver# [   28.193736] simp_le[924]:   File "/nix/store/xxg8w24n7nklwc6lcldvc3j7xwv4bfmr-python3.7-pyOpenSSL-19.0.0/lib/python3.7/site-packages/OpenSSL/SSL.py", line 1647, in _raise_ssl_error
webserver# [   28.204186] simp_le[924]:     _raise_current_error()
webserver# [   28.206762] simp_le[924]:   File "/nix/store/xxg8w24n7nklwc6lcldvc3j7xwv4bfmr-python3.7-pyOpenSSL-19.0.0/lib/python3.7/site-packages/OpenSSL/_util.py", line 54, in exception_from_error_queue
webserver# [   28.218327] simp_le[924]:     raise exception_type(errors)
webserver# [   28.222950] simp_le[924]: OpenSSL.SSL.Error: [('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')]

```

I'm a bit puzzled by the `bad record MAC`.

I thought the error could come from Python's URLlib3 ssl implementation not properly handling the certificates paths. That said, after digging into its source code, it does seems to look at the right place: https://github.com/pyca/pyopenssl/blob/master/src/OpenSSL/SSL.py#L234

I'll have a further look later tonight.

fpletz commented:
@NinjaTrappeur You've removed the override for `certifi` but that's AFAIK needed for `simp_le` to verify the ACME endpoint. That should be the issue you're experiencing.

flokli commented:
> @NinjaTrappeur You've removed the override for `certifi` but that's AFAIK needed for `simp_le` to verify the ACME endpoint. That should be the issue you're experiencing.

But why is `certifi` baking in a custom cacerts somewhere, and not using the one we provide in `/etc/ssl`?

NinjaTrappeur commented:
@flokli https://github.com/certifi/python-certifi/issues/89 :/

flokli commented:
Hm, in our case, it's `requests` pulling in `certifi` - so basically, `requests` doesn't honor the system trust store, but use `certifi` (which is basically mirroring Mozilla's trust store).Changing `request`'s behaviour nixpkgs-wide should probably involve some discussion with upstream - but it seems some people outside NixOS are struggling with this behaviour: https://incognitjoe.github.io/adding-certs-to-requests.html

However, `requests` seems to [honor the `REQUESTS_CA_BUNDLE` environment variable](https://requests.kennethreitz.org/en/master/user/advanced/#ssl-cert-verification),) (or the undocumented `CURL_CA_BUNDLE`), so passing that environment variable to `simp_le` should work.

flokli commented:
Okay, now it's only `curl` complaining about not knowing the pebble-created CA.

This should be fixable by pointing `curl` from the `testScript` to a `--capath` containing both the pebble-created CA and the system trust store.

flokli commented:
OK, got it to work - we were already passing pebble's CA to curl, but forgot to add the intermediate. Tests now pass :-)

arianvp commented:
Thanks for the hard work all

flokli commented:
@GrahamcOfBorg test acme

NinjaTrappeur commented:
Nice! Thanks a bunch folks.

fpletz reviewed (approved):
Thanks!

c0bw3b commented:
@GrahamcOfBorg build boulder certbot pebble simp_le

c0bw3b commented:
(nitpick)
`boulder` and `pebble` could be better located under  _pkgs/tools/security_  rather than  _pkgs/tools/admin_

NinjaTrappeur commented:
I'm kind of puzzled by those ofborg errors. I tried to reproduce that locally on my machine and did not manage to trigger any of those.

>  boulder and pebble could be better located under pkgs/tools/security rather than pkgs/tools/admin

`certbot` is currently stored in `pkgs/tools/admin`, I stored both `boulder` and `pebble` there out of consistency.

This PR adds both the pebble and the boulder derivation to `pkgs`. Since we do not provide any service other than the acme test for both of those, I think we should drop the one we decide not to use (be it `pebble` or `boulder`). 

I'll wait for azlig's decision - is this pebble-based test good enough or should we rollback and fix the boulder one - to drop the useless commit.

flokli commented:
@NinjaTrappeur would you mind filing a backport PR?

NinjaTrappeur commented:
Sure.

Izorkin commented:
Error build certbot in master branch:
```
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing setuptools-build-hook
Using setuptoolsBuildPhase
Using setuptoolsShellHook
Sourcing pip-install-hook
Using pipInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/i7aiyqdq7m24ki60rp7dd1f6qyzngccf-source
source root is source
setting SOURCE_DATE_EPOCH to timestamp 315619200 of file source/windows-installer/template.nsi
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
applying patch /nix/store/9zxgqy1y6mxr89bb5m7n32rns64y9qnp-0001-pebble_artifacts-hardcode-pebble-location.patch
patching file certbot-ci/certbot_integration_tests/utils/pebble_artifacts.py
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
building
Executing setuptoolsBuildPhase
Traceback (most recent call last):
  File "nix_run_setup", line 8, in <module>
    exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\\r\\n', '\\n'), __file__, 'exec'))
  File "setup.py", line 63, in <module>
    if StrictVersion(setuptools_version) >= StrictVersion('36.2'):
  File "/nix/store/zdh16dcvjw99ybam59zd2ijb6bx138j0-python3-3.7.5/lib/python3.7/distutils/version.py", line 40, in __init__
    self.parse(vstring)
  File "/nix/store/zdh16dcvjw99ybam59zd2ijb6bx138j0-python3-3.7.5/lib/python3.7/distutils/version.py", line 137, in parse
    raise ValueError("invalid version number '%s'" % vstring)
ValueError: invalid version number '41.4.0.post20191022'
```

NinjaTrappeur commented:
Reproduced on x86_64.

Not sure where this is coming from though :/ Searching for "41.4.0" in nixpkgs does not yields any result.

Looks like we'll need to bisect that.

Izorkin commented:
With setuptools v41.2.0 same error.

NinjaTrappeur commented:
It does build on top of 0a4373a46d1caede9fc24074f386d05ee08d7dd2.

We should probably bisect from there.

Izorkin commented:
In becfba9512c68b013ac62f5adb0ffc2d96c5132e - error build
in dc84a7d4e3aa42afa0d41d38ffffe98022b0f946 - normal build
https://github.com/NixOS/nixpkgs/compare/dc84a7d4e3aa42afa0d41d38ffffe98022b0f946...NixOS:becfba9512c68b013ac62f5adb0ffc2d96c5132e
No ideas where to look for a error.

FRidh commented:
The failure happens since we updated setuptools. If I am correct it is due to https://github.com/pypa/setuptools/pull/1847.

Izorkin commented:
@FRidh with setuptools v41.2.0 same error.

flokli commented:
Let's please not clutter this PR, but instead open a new issue, and properly start `git bisect`-ing the breakage.

FRidh commented:
fixed certbot in 2db400d8c844ce501f0d8009a661505297ccdf81
